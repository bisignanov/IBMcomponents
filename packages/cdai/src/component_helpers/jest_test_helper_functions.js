//
// Copyright IBM Corp. 2020, 2020
//
// This source code is licensed under the Apache-2.0 license found in the
// LICENSE file in the root directory of this source tree.
//
import React from 'react';
import { shallow, mount } from 'enzyme';

export const getJSXForComponent = (
  component,
  defaultProps = {},
  customProps = {}
) => {
  return React.createElement(component, { ...defaultProps, ...customProps });
};

export const mountComponent = (jsx, returnShallow = true) => {
  let component, componentInstance;

  if (returnShallow) {
    component = shallow(jsx);
  } else {
    component = mount(jsx);
  }

  componentInstance = component.instance();

  let unmount = () => {
    unmountComponent(component);
  };

  return {
    component,
    componentInstance,
    unmount,
  };
};

export const unmountComponent = (component) => {
  component && component.length > 0 && component.unmount();
};
