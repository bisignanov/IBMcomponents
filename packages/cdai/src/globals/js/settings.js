const settings = {
  idePrefix: 'ide',
};
module.exports = settings;
