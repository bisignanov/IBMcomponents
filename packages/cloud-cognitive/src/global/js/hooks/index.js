export { useWindowScroll } from './useWindowScroll';
export { useWindowResize } from './useWindowResize';
export { useClickOutside } from './useClickOutside';
export { useCreateComponentFocus } from './useCreateComponentFocus';
export { useCreateComponentStepChange } from './useCreateComponentStepChange';
export { usePreviousValue } from './usePreviousValue';
export { useResetCreateComponent } from './useResetCreateComponent';
export { useValidCreateStepCount } from './useValidCreateStepCount';
