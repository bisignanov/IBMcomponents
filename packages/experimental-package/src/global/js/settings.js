import featureFlags from '../../../../feature-flags/feature-flags';

const expPrefix = 'exp';

export { expPrefix };

export { featureFlags };
