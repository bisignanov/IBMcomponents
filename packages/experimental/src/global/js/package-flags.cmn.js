//
// Copyright IBM Corp. 2020, 2020
//
// This source code is licensed under the Apache-2.0 license found in the
// LICENSE file in the root directory of this source tree.
//

const defaultFlags = {
  // by default only released components are set to true
  component: {
    AboutModal: false,
    ActionBar: false,
    APIKeyDownlaoder: false,
    APIKeyModal: false,
    BreadcrumbWithOverflow: false,
    Card: false,
    ContextHeader: false,
    EmptyState: false,
    ExampleComponent: false,
    ExportModal: false,
    ImportModal: false,
    ModifiedTabs: false,
    Notifications: false,
    PageHeader: false,
    RemoveDeleteModal: false,
    SidePanel: false,
    TagSet: false,
    TearSheet: false,
  },

  // feature level flags
  feature: {
    'a-new-feature': false,
  },
};

const enableAll = (section) => {
  const result = {};
  for (const key in section) {
    result[key] = true;
  }
  return result;
};

const checkEnabled = (flags, overrides = {}) => {
  const result = {};
  for (const key in flags) {
    if (overrides[key]?.enableAll) {
      // enable all for this section
      delete overrides[key].enableAll;
      result[key] = enableAll(flags[key]);
    } else {
      result[key] = { ...flags[key], ...(overrides[key] || {}) };
    }
  }
  return result;
};

const packageFlags = () => {
  let initialized = false;
  let flags;

  // doNotUse parameter is intended for dev flags
  // doNotUse.initAgain causes flags to be refreshed which is handy in tests
  // an override section containing enableAll=true turns on everything in that section - handy for storybook
  return (overrides, doNotUse) => {
    const defaults = { ...defaultFlags };

    if (!initialized || doNotUse?.initAgain === true) {
      initialized = true;

      flags = checkEnabled(defaults, overrides);
    }

    return flags;
  };
};

const getPackageFlags = packageFlags();

module.exports = getPackageFlags;
