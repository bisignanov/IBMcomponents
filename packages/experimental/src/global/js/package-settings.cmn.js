//
// Copyright IBM Corp. 2020, 2020
//
// This source code is licensed under the Apache-2.0 license found in the
// LICENSE file in the root directory of this source tree.
//

const defaultFlags = {
  // by default only released components are set to true
  component: {
    AboutModal: false,
    ActionBar: false,
    APIKeyDownlaoder: false,
    APIKeyModal: false,
    BreadcrumbWithOverflow: false,
    Card: false,
    ContextHeader: false,
    EmptyState: false,
    ExampleComponent: false,
    ExportModal: false,
    ImportModal: false,
    ModifiedTabs: false,
    Notifications: false,
    PageActionItem: false,
    PageHeader: false,
    RemoveDeleteModal: false,
    SidePanel: false,
    TagSet: false,
    Tearsheet: false,
    TearsheetNarrow: false,
    TearsheetShell: false,
    WebTerminal: false,
  },

  // feature level flags
  feature: {
    'a-new-feature': false,
  },
};

const applyOverrides = (target, overrides = {}) => {
  for (const key in overrides) {
    if (key !== 'enableAll' && key !== 'disableAll') {
      // enable all is special handled one level up
      if (typeof target[key] === 'object') {
        if (typeof overrides[key] === 'object') {
          const disableAll = (key === 'component' || key === 'feature') && overrides[key].disableAll === true;
          const enableAll = (key === 'component' || key === 'feature') && overrides[key].enableAll === true;

          if (disableAll || enableAll) {
            // special case loop through all component or feature and enable
            for (const childKey in target[key]) {
              target[key][childKey] = enableAll;
            }
          }

          // process everything else working through branch
          applyOverrides(target[key], overrides[key]);
        } else {
          // reject replacing object with primitive
          console.warn(
            `Attempt to replace setting ${key} with primitive type rejected`
          );
        }
      } else {
        // both primitive types replace
        target[key] = overrides[key];
      }
    }
  }
};

const defaultPrefix = 'exp';

const settings = {
  prefix: defaultPrefix,
  flags: {
    component: { ...defaultFlags.component },
    feature: { ...defaultFlags.feature },
  },
  isComponentEnabled: (component, byDefault = false) => {
    const componentName = component?.name || component;
    const flags = byDefault ? defaultFlags : settings.flags;

    return flags.component[componentName];
  },
  isFeatureEnabled: (feature, byDefault = false) => {
    const flags = byDefault ? defaultFlags : settings.flags;
    return flags.feature[feature];
  },
  overrideSettings: (overrides, reset) => {
    if (reset) {
      settings.prefix = defaultPrefix;
      settings.flags = {
        component: { ...defaultFlags.component },
        feature: { ...defaultFlags.feature },
      };
    }
    applyOverrides(settings, overrides);
  },
  replaceWithCare: (replacement) => {
    console.warn(`

Presumably you know what you are doing as ou have replaced package settings:

------
${JSON.stringify(settings)}
------

with:

++++++
${JSON.stringify(replacement)}
++++++

`);

    for (const key in settings) {
      delete settings[key];
    }

    for (const key in replacement) {
      settings[key] = replacement[key];
    }
  },
};

module.exports = settings;
