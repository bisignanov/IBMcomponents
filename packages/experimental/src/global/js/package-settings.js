//
// Copyright IBM Corp. 2020, 2021
//
// This source code is licensed under the Apache-2.0 license found in the
// LICENSE file in the root directory of this source tree.
//

const defaults = {
  prefix: 'exp',
  // by default only released components are set to true
  component: {
    AboutModal: false,
    ActionBar: false,
    ActionBarItem: false,
    APIKeyDownlaoder: false,
    APIKeyModal: false,
    BreadcrumbWithOverflow: false,
    Card: false,
    ContextHeader: false,
    EmptyState: false,
    ErrorEmptyState: false,
    NoDataEmptyState: false,
    NoTagsEmptyState: false,
    NotFoundEmptyState: false,
    NotificationsEmptyState: false,
    UnauthorizedEmptyState: false,
    ExampleComponent: false,
    ExportModal: false,
    ImportModal: false,
    ModifiedTabs: false,
    Notifications: false,
    PageActionItem: false,
    PageHeader: false,
    RemoveDeleteModal: false,
    SidePanel: false,
    TagSet: false,
    Tearsheet: false,
    TearsheetNarrow: false,
    TearsheetShell: false,
    WebTerminal: false,
  },

  // feature level flags
  feature: {
    'a-new-feature': false,
  },
};

const setAll = (flags, enabled) => {
  for (const key in flags) {
    flags[key] = enabled;
  }
};

const settings = {
  // flags
  prefix: defaults.prefix,
  component: { ...defaults.component },
  feature: { ...defaults.feature },
  // methods
  isComponentEnabled: (component, byDefault = false) => {
    const componentName = component?.name || component;
    const flags = byDefault ? defaults : settings;
    return flags.component[componentName];
  },
  isFeatureEnabled: (feature, byDefault = false) => {
    const flags = byDefault ? defaults : settings;
    return flags.feature[feature];
  },
  setAllComponents: (enabled) => {
    setAll(settings.component, enabled);
  },
  setAllFeatures: (enabled) => {
    setAll(settings.feature, enabled);
  },
};

export default settings;
