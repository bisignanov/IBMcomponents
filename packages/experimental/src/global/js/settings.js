import featureFlags from '../../generated/feature-flags/feature-flags';

const expPrefix = 'exp';

export { expPrefix };

export { featureFlags };
