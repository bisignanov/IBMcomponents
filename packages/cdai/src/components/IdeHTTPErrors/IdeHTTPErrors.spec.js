//
// Copyright IBM Corp. 2020, 2020
//
// This source code is licensed under the Apache-2.0 license found in the
// LICENSE file in the root directory of this source tree.
//

import { IdeHTTPErrors } from '.';
import * as jth from '../../component_helpers/jest_test_helper_functions.js';

describe('IdeHTTPErrors', () => {
  let component, unmount;
  const mountTestComponent = (defaultProps = {}, props = {}) => {
    let test = jth.mountComponent(
      jth.getJSXForComponent(IdeHTTPErrors, defaultProps, props),
      false
    );
    component = test.component;
    unmount = test.unmount;
  };
  afterEach(() => unmount && unmount());

  // Prop based rendering tests
  it('renders when no props are defined', () => {
    mountTestComponent();
    expect(component.getElements()).toMatchSnapshot();
  });
  function rendersWhenPropIsDefined(
    propName,
    propValue = propName,
    testNameSuffix
  ) {
    let testName = `renders when the ${propName} prop is defined`;
    if (testNameSuffix) {
      testName += ` as ${testNameSuffix}`;
    }
    it(testName, () => {
      mountTestComponent(undefined, { [propName]: propValue });
      expect(component.instance().props[propName]).toBe(propValue);
      expect(component.getElements()).toMatchSnapshot();
    });
  }
  rendersWhenPropIsDefined('background', 403, '403');
  rendersWhenPropIsDefined('background', 404, '404');
  rendersWhenPropIsDefined('title');
  rendersWhenPropIsDefined('description');
  rendersWhenPropIsDefined('details');
  rendersWhenPropIsDefined('links', { text: 'text', url: 'url' }, 'an object');
  rendersWhenPropIsDefined(
    'links',
    [
      { text: 'text1', url: 'url1' },
      { text: 'text2', url: 'url2' },
    ],
    'an array'
  );

  // Validation tests
  beforeEach(() => jest.spyOn(console, 'error').mockImplementation(() => {}));
  afterEach(() => console.error.mockRestore());
  function throwsWithInvalidProp(
    propName,
    propValue = propName,
    testNameSuffix,
    errorMessageSuffix
  ) {
    let testName = `throws when ${propName} prop ${testNameSuffix}`;
    it(testName, () => {
      expect(() =>
        mountTestComponent(undefined, { [propName]: propValue })
      ).toThrowError(`IdeHTTPErrors: ${propName} prop ${errorMessageSuffix}`);
    });
  }
  throwsWithInvalidProp(
    'background',
    1,
    'is not equal to 403 or 404',
    'must be 403 or 404, got 1'
  );
  throwsWithInvalidProp(
    'title',
    1,
    'is not type string',
    'must be type string, got number'
  );
  throwsWithInvalidProp(
    'description',
    1,
    'is not type string',
    'must be type string, got number'
  );
  throwsWithInvalidProp(
    'details',
    1,
    'is not type string',
    'must be type string, got number'
  );
  describe('links prop', () => {
    throwsWithInvalidProp(
      'links',
      1,
      'is not an object or array',
      'must be an object or array of objects with text+url props, got 1'
    );
    function throwsWithInvalidNestedProp(
      propName,
      propValue,
      nestedPropName,
      testNameSuffix,
      errorMessageSuffix
    ) {
      let propNameSuffix = '.';
      if (Array.isArray(propValue)) {
        propNameSuffix = '[n].';
      }
      let testName = `throws when ${propName}${propNameSuffix}${nestedPropName} prop is ${testNameSuffix}`;
      it(testName, () => {
        expect(() =>
          mountTestComponent(undefined, { [propName]: propValue })
        ).toThrowError(
          `IdeHTTPErrors: ${propName}.${nestedPropName} prop ${errorMessageSuffix}`
        );
      });
    }
    describe('defined as an object', () => {
      throwsWithInvalidNestedProp(
        'links',
        { text: 1, url: '' },
        'text',
        'is not type string',
        'must be type string, got number'
      );
      throwsWithInvalidNestedProp(
        'links',
        { text: '', url: 1 },
        'url',
        'is not type string',
        'must be type string, got number'
      );
    });
    describe('defined as an array', () => {
      throwsWithInvalidProp(
        'links',
        [],
        'defined as an array does not contain any items',
        'defined as an array must contain at least 1 item'
      );
      throwsWithInvalidNestedProp(
        'links',
        [{ text: 1, url: '' }],
        'text',
        'is not type string',
        'must be type string, got number'
      );
      throwsWithInvalidNestedProp(
        'links',
        [{ text: '', url: 1 }],
        'url',
        'is not type string',
        'must be type string, got number'
      );
    });
  });

  // Coverage completion tests
  describe('renderBackground method', () => {
    it('returns null when inline and viewport width is greater than or equal to medium', () => {
      mountTestComponent();
      component.setState({ width: 672 });
      expect(component.instance().renderBackground(true)).toEqual(null);
    });
  });
  describe('renderSection method', () => {
    it('returns null when sectionClassName is undefined', () => {
      mountTestComponent();
      expect(component.instance().renderSection()).toEqual(null);
    });
  });
});
