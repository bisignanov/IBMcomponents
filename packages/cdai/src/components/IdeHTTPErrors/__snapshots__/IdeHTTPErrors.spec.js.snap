// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IdeHTTPErrors renders when no props are defined 1`] = `
[
  <IdeHTTPErrors
    description="CHANGE ME"
    title="CHANGE ME"
  />,
]
`;

exports[`IdeHTTPErrors renders when the background prop is defined as 403 1`] = `
[
  <IdeHTTPErrors
    background={403}
    description="CHANGE ME"
    title="CHANGE ME"
  />,
]
`;

exports[`IdeHTTPErrors renders when the background prop is defined as 404 1`] = `
[
  <IdeHTTPErrors
    background={404}
    description="CHANGE ME"
    title="CHANGE ME"
  />,
]
`;

exports[`IdeHTTPErrors renders when the description prop is defined 1`] = `
[
  <IdeHTTPErrors
    description="description"
    title="CHANGE ME"
  />,
]
`;

exports[`IdeHTTPErrors renders when the details prop is defined 1`] = `
[
  <IdeHTTPErrors
    description="CHANGE ME"
    details="details"
    title="CHANGE ME"
  />,
]
`;

exports[`IdeHTTPErrors renders when the links prop is defined as an array 1`] = `
[
  <IdeHTTPErrors
    description="CHANGE ME"
    links={
      [
        {
          "text": "text1",
          "url": "url1",
        },
        {
          "text": "text2",
          "url": "url2",
        },
      ]
    }
    title="CHANGE ME"
  />,
]
`;

exports[`IdeHTTPErrors renders when the links prop is defined as an object 1`] = `
[
  <IdeHTTPErrors
    description="CHANGE ME"
    links={
      {
        "text": "text",
        "url": "url",
      }
    }
    title="CHANGE ME"
  />,
]
`;

exports[`IdeHTTPErrors renders when the title prop is defined 1`] = `
[
  <IdeHTTPErrors
    description="CHANGE ME"
    title="title"
  />,
]
`;
